cpu/esp_common/include/irq_arch_common.h

```diff
@@ -67,6 +67,79 @@ extern volatile uint32_t irq_interrupt_nesting;
 #define critical_exit_var(m)    irq_restore(m)
 /** @} */
 
+/**
+ * @brief Disable all maskable interrupts
+ */
+__attribute__((always_inline)) static inline unsigned int irq_disable(void)
+{
+    uint32_t state;
+
+    /* read and set interrupt level (RSIL) */
+    __asm__ volatile (
+        /* this will block IRQs with prio <= intlevel */
+        "rsil %[state], %[intlevel]"
+        : [state]       "=a" (state)
+        : [intlevel]    "I"(15) /* 15 is highest prio, blocking all IRQs */
+        : "memory"
+    );
+    /* mask out everything else of the PS register that do not belong to
+       interrupt level (bits 3..0) */
+    state &= 0xf;
+
+    return state;
+}
+
+/**
+ * @brief Enable all maskable interrupts
+ */
+__attribute__((always_inline)) static inline unsigned int irq_enable(void)
+{
+    uint32_t state;
+
+    /* read and set interrupt level (RSIL) */
+    __asm__ volatile (
+        "rsil %[state], %[intlevel]"
+        : [state]       "=a"(state)
+        : [intlevel]    "I"(0) /* 0 is lowest prio, unblocking all IRQs */
+        : "memory"
+    );
+    /* mask out everything else of the PS register that do not belong to
+       interrupt level (bits 3..0) */
+    state &= 0xf;
+
+    return state;
+}
+
+/**
+ * @brief Restore the state of the IRQ flags
+ */
+__attribute__((always_inline)) static inline void irq_restore(unsigned int state)
+{
+    /* the rsil instruction requires an constant immediate, thus, we cannot
+     * pass the state parameters via register. Instead, we rely on the fact
+     * that RIOT does not support different IRQ levels, but instead only enables
+     * or disables *all* IRQs. So if state is 0 (all IRQs enabled), we will
+     * do so using rsil, otherwise we don't touch the interrupt level
+     */
+    if (!state) {
```

Comment:
With this implementation, you can't restore interrupt levels that are greater than 0. This situation could occur if `irq_disable` is called while the `PS.INTLEVEL` register is already greater than 0 for some reason, e.g., because the low level interrupts have already been disabled by SDK code that we do not control.

Example: `periph/timer` uses interrupt 19 (level 2). Let's suppose that the SDK code has only disabled low level interrupts (level 1) what is really possible. Then the `PS.INTLEVEL` register is 1. If the timer then triggers an interrupt, RIOT code is executed. That is, `irq_disable` is called on entry with `PS.INTLEVEL=1` and `irq_restore` is called on exit. However, `irq_disable` wouldn't restore the old interrupt level and all interrupts would remain disabled.

I don't know how probable it is, but it can happen in theory. If it did happen, we would sporadically get errors that we all "love" so much because they are not reproducible. As this change could have a major impact on stability, I would suggest not making this change. I would sacrifice performance in favor of stability.